### ISSUE TITLE
caddy reload does not return to the command line when finished

### ISSUE THREAD

**Issue by ale-rinaldi** [caddyserver/caddy#3540]
Hello,

I'm experiencing this same behaviour as issue #3341 on latest v2.1.1 release (didn't happen on 2.0.0).
```
caddy@myserver:~$ caddy version
v2.1.1 h1:X9k1+ehZPYYrSqBvf/ocUgdLSRIuiNiMo7CvyGUQKeA=
```

This is what happens:

```
caddy@myserver:~$ caddy reload --config /etc/caddy/Caddyfile
2020/07/01 10:55:14.457	INFO	using provided configuration	{"config_file": "/etc/caddy/Caddyfile", "config_adapter": ""}

```

and stays like this until I hit CTRL-C. As a consequence, even `sudo systemctl reload caddy` is failing for timeout. However, the configuration seems to get reloaded properly.

On Caddy side, I only see this log being outputted:
```
{"level":"info","ts":1593613252.249103,"logger":"admin.api","msg":"received request","method":"POST","host":"localhost:2019","uri":"/load","remote_addr":"127.0.0.1:43840","headers":{"Accept-Encoding":["gzip"],"Content-Length":["2503"],"Content-Type":["application/json"],"Origin":["localhost:2019"],"User-Agent":["Go-http-client/1.1"]}}
```
Nothing else is reported about the reload.

My Caddyfile is quite simple (I just replaced the domain since it's a customer's one):

```
monitor.mydomain.com {
    route /netdata/* {
        uri strip_prefix /netdata
        reverse_proxy 127.0.0.1:19999
    }
    redir /netdata /netdata/

    basicauth /netdata/* {
        bob JDJhJDEwJEVCNmdaNEg2Ti5iejRMYkF3MFZhZ3VtV3E1SzBWZEZ5Q3VWc0tzOEJwZE9TaFlZdEVkZDhX
    }
}

http://sub1.mydomain.com,
https://sub1.mydomain.com {
    reverse_proxy 127.0.0.1:8000
}

http://sub2.mydomain.com,
https://sub2.mydomain.com {
    reverse_proxy 127.0.0.1:8001
}

http://sub3.mydomain.com,
https://sub3.mydomain.com {
    reverse_proxy 127.0.0.1:8002
}

http://sub4.mydomain.com,
https://sub4.mydomain.com {
    reverse_proxy 127.0.0.1:8003
}
```

**Comment by mholt** [comment]
Thanks for opening an issue! We'll look into this.

It's not immediately clear to me what is going on, so I'll need your help to understand it better.

Ideally, we need to be able to reproduce the bug _in the most minimal way possible_. This allows us to write regression tests to verify the fix is working. If we can't reproduce it, then you'll have to test our changes for us until it's fixed -- and then we can't add test cases, either.

I've attached a template below that will help make this easier and faster! **This will require some effort on your part -- please understand that we will be dedicating time to fix the bug you are reporting if you can just help us understand it and reproduce it easily.**

This template will ask for some information you've already provided; that's OK, just fill it out the best you can. :+1: I've also included some helpful tips below the template. Feel free to let me know if you have any questions!

Thank you again for your report, we look forward to resolving it!

## Template

<pre><code>## 1. Environment

### 1a. Operating system and version

```
paste here
```


### 1b. Caddy version (run `caddy version` or paste commit SHA)

```
paste here
```


### 1c. Go version (if building Caddy from source; run `go version`)

```
paste here
```


## 2. Description

### 2a. What happens (briefly explain what is wrong)




### 2b. Why it's a bug (if it's not obvious)




### 2c. Log output

```
paste terminal output or logs here
```



### 2d. Workaround(s)




### 2e. Relevant links




## 3. Tutorial (minimal steps to reproduce the bug)




</code></pre>

## Helpful tips

1. **Environment**: Please fill out your OS and Caddy versions, even if you don't think they are relevant. (They are _always_ relevant.) If you built Caddy from source, provide the commit SHA and specify your exact Go version.

2. **Description:** Describe at a high level what the bug is. What happens? Why is it a bug? Not all bugs are obvious, so convince readers that it's actually a bug.
	- **2c) Log output:** Paste terminal output and/or complete logs in a code block. [DO NOT REDACT INFORMATION except for credentials.](https://caddy.community/t/how-to-get-help-with-caddy-more-effectively/5222)
	- **2d) Workaround:** What are you doing to work around the problem in the meantime? This can help others who encounter the same problem, until we implement a fix.
	- **2e) Relevant links:** Please link to any related issues, pull requests, docs, and/or discussion. This can add crucial context to your report.

3. **Tutorial:** What are the _minimum required specific steps_ someone needs to take in order to experience the same bug? Your goal here is to make sure that anyone else can have the same experience with the bug as you do. You are writing a tutorial, so make sure to carry it out yourself before posting it. Please:
	- Start with an empty config. Add _only_ the lines/parameters that are _absolutely required_ to reproduce the bug.
	- Do not run Caddy inside containers.
	- Run Caddy manually in your terminal; do not use systemd or other init systems.
	- If making HTTP requests, avoid web browsers. Use a simpler HTTP client instead, like `curl`.
	- Do not redact any information from your config (except credentials). Domain names are public knowledge and [often necessary for quick resolution of an issue](https://caddy.community/t/how-to-get-help-with-caddy-more-effectively/5222)!
	- Note that ignoring this advice may result in delays, or even in your issue being closed. 😞 Only actionable issues are kept open, and if there is not enough information or clarity to reproduce the bug, then the report is not actionable.

Example of a tutorial:

<blockquote>Create a config file:

```
{ ... }
```

Open terminal and run Caddy:

```
$ caddy ...
```

Make an HTTP request:

```
$ curl ...
```

Notice that the result is ___ but it should be ___.</blockquote>

**Comment by mholt** [comment]
Closing due to inactivity / lack of info.

**Comment by sixtus** [comment]
fyi: this happens randomly to me too.

**Comment by francislavoie** [comment]
@sixtus that's not helpful, we can't do anything without a reproduce case.

**Comment by sixtus** [comment]
I found the solution, it's a missing grace_period - if there is a connection that makes the reload hang beyond the timeout of systemd, reload is effectively ignored.

**Comment by sixtus** [comment]
ok, so grace_period is related, but not in the code path that's responsible for my trouble. My neighbor was so nice to start the debugger, this is his wip hotfix: https://github.com/burton1982/caddy/pull/1

**Comment by mholt** [comment]
But why is `Stop()` hanging? :thinking:

**Comment by m3nu** [comment]
Had the same issue as described here: Reloads timing out or taking more than  minute. The linked patch resolved it. Might be worth investigating, @mholt 

My Caddyfile is structurally similar to OP's: many subdomains proxying to different ports. So might be related to that. Happy to help debug further.

Thanks to @sixtus and @burton1982 for the debugging.

**Comment by m3nu** [comment]
With the patch applied, I'm seeing those errors: `[ERROR] stop tls: 0xc000e0e1e0`

Would this mean there are errors while stopping the `tls` app?

**Comment by mholt** [comment]
If `Stop()` is indeed blocking, then that patch leaks goroutines. New bug that just hides the actual problem.

**Comment by m3nu** [comment]
Thanks for looking into it. 🙏 Of course best to resolve the actual problem. So we need to look in `tls.stop()` (pseudocode) and the reason why it doesn't stop properly?

**Comment by burton1982** [comment]
:-) Just to be clear. This is why I told my neighbor to call my patch 'dirty workaround'. It is clear that this would not solve anything but highlight the issue.

**Comment by m3nu** [comment]
Checked my logs again. Here what's failing to stop: (the log lines near the patch)

```
2022/01/26 06:25:15 [ERROR] stop tls: 0xc001026480
2022/01/26 06:25:53 [ERROR] stop tls: 0xc000e0e1e0
2022/01/26 06:34:26 [ERROR] stop tls: 0xc000d83e60
2022/01/26 06:43:48 [ERROR] stop tls: 0xc00118b4a0
2022/01/26 06:57:58 [ERROR] timeout stopping http
2022/01/26 06:59:03 [ERROR] timeout stopping http
2022/01/26 07:02:08 [ERROR] timeout stopping http
```

**Comment by m3nu** [comment]
Can we reopen this, since it's clearly an ongoing issue? My Caddyfile is similar to OP's. And I'm happy to set up a test server to trigger the issue.

**Comment by m3nu** [comment]
The config alone isn't enough to reproduce. Might be related to active connections (websocket?) at the time of the reload.

**Comment by mholt** [comment]
@m3nu And this is _with_ a grace period set?

**Comment by m3nu** [comment]
I'll observe and try the released vs. patched version. 🤞 But this option sounds promising.

**Comment by m3nu** [comment]
I believe this does the trick. Sorry for the noise here. With grace period set, I get this, but rest works and reloads `[ERROR] stop http: context deadline exceeded`.

For future reference if anyone finds this: Add this to your `Caddyfile` (global config):

```
{
    grace_period 5s
}
```

**Comment by mholt** [comment]
Yeah, I'm pretty sure that's why the issue was closed, as @sixtus said earlier:

> I found the solution, it's a missing grace_period 

and this is expected; if a connection stays active forever (like websockets, or a really slow, big download or upload), then Caddy can't gracefully close it without waiting for it to be idle. Hence the grace period and, if reached, the expected "context deadline exceeded" error which basically means we stopped waiting for it to stop and just forced it.

But we don't want to enable this by default as it can break a lot of legitimate connections too, and the right grace period for everyone is different.
